### PR TITLE
Update deploy action for release 2.x

### PR DESCRIPTION
**Pull request by nainkunal933**
**Issue #:**
Deploy script should handle multiple branches and metrics should not send all metrics to the same namespace. 
**Description of changes:**
The deploy action needs to be updated to deploy demos against 2.x branch along with the latest branch. Metrics need to account for previous major release tests. 
**Testing:**

*Can these tested using a demo application? Please provide reproducible step-by-step instructions.*
No. These are integration test changes which were tested locally with KITE. 

**Checklist:**

1. Have you successfully run `npm run build:release` locally?
Yes

2. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
No

3. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/deploy.yml
Comment (LeviSklaroff) [inline .github/workflows/deploy.yml:28]: I believe these are redundant as the default value should be the SHA of the commit triggering this workflow. https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request

Comment (xuesichao) [inline .github/workflows/deploy.yml:28]: Could tell me which branch is this `ref_name` referring to? and where we could access this variable?

File: integration/js/utils/SdkBaseTest.js
Comment (xuesichao) [inline integration/js/utils/SdkBaseTest.js:168]: Shall we use a global variable `this.isPreviousMajorVersion` to replace `this.paylaod?.isPreviousMajorVersion`?

File: .github/workflows/deploy.yml
Comment (nainkunal933) [reply .github/workflows/deploy.yml:28]: The `ref_name` will change based on which branch this workflow was triggered on. You can find more information about it under GitHub context here: https://docs.github.com/en/actions/learn-github-actions/contexts#github-context

Documentation:


github.ref_name | string | The branch or tag name that triggered the workflow run.
-- | -- | --

File: integration/js/utils/SdkBaseTest.js
Comment (nainkunal933) [reply integration/js/utils/SdkBaseTest.js:168]: We can use a global variable for `isPreviousMajorVersion` but there is no real benefit of doing it. 

This variable is being sent to KITE via the `payload` property. Some tests will have this property set to true while others will not have this property. We can make it a global property and set it to false by default and overwrite it whenever we do get it from the payload but not sure if its worth it.

File: .github/workflows/deploy.yml
Comment (nainkunal933) [reply .github/workflows/deploy.yml:28]: The default value would be the branch name that triggered the workflow so I do not think this is redundant.

Comment (xuesichao) [reply .github/workflows/deploy.yml:28]: Yeah according to the [doc](https://github.com/actions/checkout#usage):
> 
```
- uses: actions/checkout@v2
  with:
    # Repository name with owner. For example, actions/checkout
    # Default: ${{ github.repository }}
    repository: ''

    # The branch, tag or SHA to checkout. When checking out the repository that
    # triggered a workflow, this defaults to the reference or SHA for that event.
    # Otherwise, uses the default branch.
    ref: ''
    ...
```

It is default to the main branch and will change based on which event triggers this workflow. Do we need to declare it explictly?

File: integration/js/utils/SdkBaseTest.js
Comment (xuesichao) [reply integration/js/utils/SdkBaseTest.js:168]: I think currently we are assigning the payload data to global variables in the constructor. If there's no specific reason of not doing this, I would prefer to follow the convention. There's no specific benefit of doing this in this case. But it is a good practice to follow. The constructor could helps ensure the variable is in the good states before we use it. 
Although `this.payload?.isPreviousMajorVersion)` could handle the same logic here, but constructor could better manager the states.

https://github.com/aws/amazon-chime-sdk-js/blob/fca27b4f6e45160b7fbbde9115b16c826ed472bf/integration/js/utils/SdkBaseTest.js#L19-L25

**Comment by xuesichao** [comment]
Split this PR into two as discussed earlier:
PR to update the metric reporting for release 2.x: https://github.com/aws/amazon-chime-sdk-js/pull/1987/files

**Comment by LeviSklaroff** [comment]
Successful run: https://github.com/aws/amazon-chime-sdk-js/runs/5060868746?check_suite_focus=true
Was ran by editing required workflow params to run on this branch after a pull_request

**Comment by LeviSklaroff** [comment]
> Successful run: https://github.com/aws/amazon-chime-sdk-js/runs/5060868746?check_suite_focus=true Was ran by editing required workflow params to run on this branch after a pull_request

Correction, some jobs failed. Will investigate.

**Comment by LeviSklaroff** [comment]
https://github.com/aws/amazon-chime-sdk-js/runs/5071164404?check_suite_focus=true last step is expected to fail at this time

File: .github/workflows/deploy.yml
Comment (ltrung) [inline .github/workflows/deploy.yml:60]: This is nice but does that mean I have to update the demo URLs in canary config for prev major version (from 2.x to 3.x) everytime we have a new major version?
It is not as nice but should we just hard code to prev-major version so we don't have to update config.

File: .github/workflows/deploy.yml
Comment (ltrung) [inline .github/workflows/deploy.yml:63]: Nice way doing this :))

**Comment by ltrung** [comment]
@LeviSklaroff Can we rename the PR title and description to remove all the metric part ?

File: .github/workflows/deploy.yml
Comment (LeviSklaroff) [reply .github/workflows/deploy.yml:60]: That's how it would work right now, but we would have to update the configs anyways for their meta data anyways, so that it's not updating 2x metrics when we're on 3x, this also allows us the option to support more then just the last version, but I do see your point and I wouldn't say I disagree with just hard coding it. We could make it pretty easy to update though by just making it a single variable.

File: .github/workflows/deploy.yml
Comment (LeviSklaroff) [reply .github/workflows/deploy.yml:63]: Thanks!

**Comment by LeviSklaroff** [comment]
> @LeviSklaroff Can we rename the PR title and description to remove all the metric part ?

Oops lol, done.

File: .github/workflows/deploy.yml
Comment (ltrung) [reply .github/workflows/deploy.yml:60]: I think for metrics it is already prev_version or something hard-coded.  There is no 2.x metrics :D

Comment (LeviSklaroff) [reply .github/workflows/deploy.yml:60]: Ah ok, then I'll make it prev-major-version if we ever want to update it to be version number won't be a hard change.

**Comment by LeviSklaroff** [comment]
Run after changing to hard coded 'prev-major-version': https://github.com/aws/amazon-chime-sdk-js/runs/5072728101?check_suite_focus=true

**Comment by LeviSklaroff** [comment]
> Run after changing to hard coded 'prev-major-version': https://github.com/aws/amazon-chime-sdk-js/runs/5072728101?check_suite_focus=true

Looks like we reached the bucket character limit, I'll change it to be "-prev-major"  instead

File: script/deploy-canary-demo
Comment (xuesichao) [inline script/deploy-canary-demo:3]: Could we rename this `versionInfix` to `version`, I think it is not `Infix` in this context.

Comment (xuesichao) [reply script/deploy-canary-demo:3]: Just want to clarify a little bit. 
If I'm right,`${1+$1}` returns `$1` only when `$1` exists.
So the goal you want to achieve is:
* By default the version should be empty for latest Canary
* If there's a version number, append it at the end of the demo name

File: script/deploy-canary-demo
Comment (xuesichao) [inline script/deploy-canary-demo:53]: nit: add new line end of the file

Comment (LeviSklaroff) [reply script/deploy-canary-demo]: Added

File: script/deploy-canary-demo
Comment (LeviSklaroff) [reply script/deploy-canary-demo]: Sure, changed

Comment (LeviSklaroff) [reply script/deploy-canary-demo]: > Just want to clarify a little bit. If I'm right,`${1+$1}` returns `$1` only when `$1` exists. So the goal you want to achieve is:
> 
> * By default the version should be empty for latest Canary
> * If there's a version number, append it at the end of the demo name

That's correct

**Review by xuesichao — DISMISSED** [review]
Looks good to me, but you might want to get approval from Trung as well before merging.

File: script/deploy-canary-demo
Comment (ltrung) [inline script/deploy-canary-demo:3]: Kinda like the infix or suffix name earlier. This prev-major is not a version :p.
Also I never remember any bash syntax but I assume this is a just a fancy way of saying if $1 is set then version = $1? Is it any different from just version = $1?

Comment (LeviSklaroff) [reply script/deploy-canary-demo:3]: How about `canarySufix`? Also it is effectively version=$1, but the main difference is that when I tested giving no param to the script with version=$1 it gave a error because 1 was never set, but this syntax is safer because it effectively checks if $1 exists before setting it to $1, effectively equates to this:
`
if($1)
  version=$1
else
  version=nothing
`

Comment (xuesichao) [reply script/deploy-canary-demo:3]: `canarySuffix` looks good to me.

Comment (xuesichao) [reply script/deploy-canary-demo:3]: oh right, could you also add a comment for this line?
> if($1) version=$1 else version=nothing
this is good enough

Comment (ltrung) [reply script/deploy-canary-demo:3]: Yeah canarySuffix sounds good enough

Comment (LeviSklaroff) [reply script/deploy-canary-demo]: Updated, although deploying for previous version won't work until our s3 bucket limit is increased, but based on output there isn't anything wrong with the code itself and we can merge. https://github.com/aws/amazon-chime-sdk-js/runs/5073025117?check_suite_focus=true

**Review by ltrung — APPROVED** [review]

**Pull request merged**
